### PR TITLE
Refactor parts of SignUrlRequest to base class.

### PR DESCRIPTION
**Pull request by coryan**
This will be used to create a V4SignUrlRequest, which shares many, but
not all, the properties of a V2 signed url. I think renaming
SignUrlRequest to V2SignUrlRequest should happen in a future PR too.

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/2121)

### CODE REVIEW

File: google/cloud/storage/internal/signed_url_requests.cc [hunk 1]
```diff
@@ -24,13 +24,24 @@ namespace storage {
...
+  }
+  auto kv = o.value();
+  // Normalize the header, they are not case sensitive.
+  std::transform(kv.first.begin(), kv.first.end(), kv.first.begin(),
+                 [](unsigned char c) { return std::tolower(c); });
+  auto res = extension_headers_.insert(kv);
+  if (!res.second) {
+    // The element already exists, we need to append:
+    res.first->second.push_back(',');
+    res.first->second.append(o.value().second);
```
Comment (devbww) [inline google/cloud/storage/internal/signed_url_requests.cc]: `kv.second`?

File: google/cloud/storage/internal/signed_url_requests.h [hunk 2]
```diff
@@ -27,25 +27,69 @@ namespace cloud {
...
+  std::string sub_resource_;
+  std::map<std::string, std::string> extension_headers_;
+  std::multimap<std::string, std::string> query_parameters_;
+};
+
 /**
  * Requests the Google Cloud Storage service account for a project.
  */
-class SignUrlRequest {
+class SignUrlRequest : private SignUrlRequestCommon {
```
Comment (devbww) [inline google/cloud/storage/internal/signed_url_requests.h]: Q. Do we have any style preference for composition over private inheritance?

File: google/cloud/storage/internal/signed_url_requests.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/internal/signed_url_requests.cc]: Doh, bad merge, fixed.

File: google/cloud/storage/internal/signed_url_requests.h (hunk 2 above)
Comment (coryan) [reply google/cloud/storage/internal/signed_url_requests.h]: Not that I know of.  This seems like a decent use for private inheritance, but obviously we can change it (at a cost of more code I believe).

**Comment by coryan** [comment]
PTAL.

File: google/cloud/storage/internal/signed_url_requests.h (hunk 2 above)
Comment (devjgm) [reply google/cloud/storage/internal/signed_url_requests.h]: FWIW: https://google.github.io/styleguide/cppguide.html#Inheritance

TL;DR: Inheritance should be public. Prefer composition over inheritance.

Comment (coryan) [reply google/cloud/storage/internal/signed_url_requests.h]: Private inheritance **is** composition in my book, but meh.  Should I change this?

Comment (devjgm) [reply google/cloud/storage/internal/signed_url_requests.h]: I personally won't request a change in this PR. But perhaps we should talk about this a bit more separately so that as a team we can come to some agreement one way or the other on the pattern to use going forward.

Also, I agree that private inheritance is very similar to composition, but I don't view them as equivalent. All inheritance, including private, creates a tighter coupling between classes, for one example, because the subclass has access to the parent's protected members. Composition provides a looser coupling between classes because each class can only access the public interface of the other.... and maybe I've just been swayed too much by Scott Meyers's advice on private inheritance :-)

Comment (coryan) [reply google/cloud/storage/internal/signed_url_requests.h]: Made the change, not worth worrying about it.

**Comment by coryan** [comment]
PTAL.
